language-plutus-core/test/Language/PlutusCore/Gen/Type.hs

```diff
@@ -0,0 +1,261 @@
+{-|
+Description: PLC Syntax, typechecker,semantics property based testing.
+
+This file contains
+1. A duplicate of the Plutus Core Abstract Syntax (types and terms)
+2. A kind checker and a type checker
+3. Reduction semantics for types
+-}
+
+{-# LANGUAGE DeriveDataTypeable        #-}
+{-# LANGUAGE FlexibleInstances         #-}
+{-# LANGUAGE LambdaCase                #-}
+{-# LANGUAGE RecordWildCards           #-}
+{-# LANGUAGE ScopedTypeVariables       #-}
+{-# LANGUAGE TemplateHaskell           #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE ExistentialQuantification #-}
+
+module Language.PlutusCore.Gen.Type
+  ( TypeBuiltinG (..)
+  , TypeG (..)
+  , ClosedTypeG
+  , toClosedType
+  , checkClosedTypeG
+  , normalizeTypeG
+  , GenError (..)
+  , ErrorP (..)
+  ) where
+
+import           Control.Enumerable
+import           Control.Monad.Except
+import           Data.Coolean
+import qualified Data.Text                      as Text
+import           Language.PlutusCore
+import           Language.PlutusCore.Gen.Common
+
+-- * Enumerating builtin types
+
+data TypeBuiltinG
+  = TyByteStringG
+  | TyIntegerG
+  | TyStringG
+  deriving (Typeable, Eq, Show)
+
+deriveEnumerable ''TypeBuiltinG
+
+-- |Convert generated builtin types to Plutus builtin types.
+toTypeBuiltin :: TypeBuiltinG -> Some (TypeIn DefaultUni)
+toTypeBuiltin TyByteStringG = Some (TypeIn DefaultUniByteString)
+toTypeBuiltin TyIntegerG    = Some (TypeIn DefaultUniInteger)
+toTypeBuiltin TyStringG     = Some (TypeIn DefaultUniString)
+
+
+-- * Enumerating types
+
+data TypeG n
+  = TyVarG n
+  | TyFunG (TypeG n) (TypeG n)
+  | TyIFixG (TypeG n) (Kind ()) (TypeG n)
+  | TyForallG (Kind ()) (TypeG (S n))
+  | TyBuiltinG TypeBuiltinG
+  | TyLamG (TypeG (S n))
+  | TyAppG (TypeG n) (TypeG n) (Kind ())
+  deriving (Typeable, Eq, Show)
+
+deriveEnumerable ''Kind
+
+deriveEnumerable ''TypeG
```

Comment:
Do we naively enumerate type-level lambda terms for testing purposes? This is [known to produce extremely boring results](http://okmij.org/ftp/Haskell/AlgorithmsH1.html#random-term) even in untyped lambda calculus. In our setting enumerating types is going to be even worse, because most of them are not going to be well-kinded.

Reply:
> I don't think so. I think the approach is to generate well-kinded types.

From what I can say by reading the code, we search through the space of all types up to a certain depth, filter out ill-kinded types and check various properties for remaining ones. I don't know how clever `lazy-search` is, but given that we have the following ingredients:

- a derived `Enumerable` instance
- an invariant checked by a custom function (`checkTypeG`)
- a library handling the general case

I'm sceptical that we produce quality test data and not spend the majority of time filtering out garbage. Maybe I'm wrong, but I do think that collecting and presenting some statistics would be extremely useful.

Reply:
@effectfully We're *lazily* walking the search tree with all types (both well- and ill-kinded) at the leaf nodes, and partial types (i.e. types with holes) at each branching node. We prune any branch as soon as the partial type is ill-kinded.

Reply:
> We're lazily walking the search tree with all types (both well- and ill-kinded) at the leaf nodes, and partial types (i.e. types with holes) at each branching node

Thanks, I now understand how this whole thing works.

> We prune any branch as soon as the partial type is ill-kinded.

This is what I mean by "discarding". We're still running into ill-kinded types and discard them, even though we do that for all types with a given prefix only once. Can we at least print the value stored in the `Left` here:

```haskell
  result <- ctrex' O depth (toTyPropG typrop k)
  case result of
    Left  _   -> return ()
```

to see how many well-kinded types were generated?

Reply:
```
  kind checker for generated types is sound:         OK (0.68s)
    2900 types generated
  normalization preserves kinds:                     OK (4.32s)
    2900 types generated
  normalization for generated types is sound:        OK (1.76s)
    2900 types generated
```